libr/include/r_vector.h

```diff
@@ -23,25 +23,25 @@ typedef struct r_vector_t {
 	size_t len;
 	size_t capacity;
 	size_t elem_size;
+	RVectorFree free;
+	void *free_user;
 } RVector;
 
-typedef struct r_pvector_t {
-	RVector v;
-	RPVectorFree free;
-} RPVector;
+// RPVector directly wraps RVector for type safety
```

Comment:
I'm really not convinced by this "type safety". I think pvector is not necessary at all. In any case, you have to cast the result of `void *` (or `void **`) to your exact type, so you are already losing safety I think.

Reply:
Yeah you can clearly use a regular RVector to store pointers. RPVector ist primarily for convenience and especially also so it is more consistent with how RList works. Otherwise you always need to have a variable that your pointer is stored in and never forget that you can't just pass the pointer.

Also, vector always has to use memcpy to assign values, because the element size can be anything. That's not a big deal but imo it is nicer to use a regular assignment if possible, which pvector does.

Reply:
I agree, pvector seems more like rlist. Then why can't we treat everything in that way? RList/RTree/RGraph/etc. do not have different APIs when they deal with pointers. Why do non-pointers need a special API?

Reply:
They don't have special apis for non-pointers, because they simply only deal with pointers. Those are all linked-list-like structures, they need to allocate each element on the heap separately anyway, so every element is always a pointer.

But for vectors it would be a waste not to have the possibility to for example have a flat array of a certain struct instead of an array of pointers to such that struct. You could even use a vector as a dynamically resizeable string if you wanted.

I know it is not perfect with vector and pvector. With C++ templates for example I would just do something like `std::vector<Something>` vs. `std::vector<Something *>`, but we just don't have that in C.

But to me it is just important to have both possibilities: A flat vector of anything or a vector of pointers that is not a complete pain to use and as consistent as possible with RList, etc.

Reply:
I agree it's important to have both possibilities. I'm trying to understand, because maybe there's something I'm missing. Why a `vector of pointers` can't be treated just as a `vector of anything`? When you want to store, for example, a flat vector of a structure, you are assuming you have some "data" of size `sizeof(struct X)`. Why for pointers can't you just assume you have some "data" of size `sizeof(struct X *)`?

Reply:
It can, absolutely. That's basically what is happening internally when you use pvector. But for example a regular free function would take `struct X *` for example. But vector simply can not know that you are storing pointers, so the only thing it can pass to the free function is `struct X **` and thus you always have to make a custom free function that dereferences the ** and passes that to your real free function for `struct X *`. Same goes for inserting. You can't give it `struct X *` directly, because for vector it could be anything of any size.

So if we say we only want the regular vector, then the api for it must stay like that, which is a pita to use with pointers and inconsistent with RList.
It's a tradeoff. But in my opinion it's better to have two different types for explicitly specified purposes than one that's annoying to use.

Reply:
Ok I understand now.

Though I don't see a lot of problems. Examples:
```
struct mystruct {
   char *heapallocated;
};

void mystruct_free1(struct mystruct *obj) {
   free(obj->heapallocated);
}

void mystruct_free2(struct mystruct **obj) {
   struct mystruct *o = *obj;
   mystructure_free(o);
}

// this is the free function defined for mystruct (e.g. r_flag_free, r_bin_free, etc.)
void mystruct_free(struct mystruct *obj) {
   free(obj->heapallocated);
   free(obj);
}

// vector with raw structures
r_vector_new (sizeof(struct mystruct), (RVectorFree)mystruct_free1);

// vector with pointer to structures
r_vector_new (sizeof(struct mystruct *), (RVectorFree)mystruct_free2);
```

Also, note that in the "flat raw structures" you can't reuse the common "free" method that we usually have for all our heap-allcoated objects (e.g. `r_anal_ref_free`, `r_hash_free`, etc.). So it's in general "easier" and more efficient to just store pointers.

What do you think?

@radare opinions?

Reply:
Yeah it depends on what one defines as a problem. I just think always having to define a function like `mystruct_free2` is something that would better be done once and reused. Same for assigning. And again, that way pvector ist just more consistent with RList.

Reply:
Ok I do see. I guess it's fine with two different APIs then. Maybe just explain it a bit better in the r_vector.h file, inside some comments ;)